File: apps/console/tests/unit/rbac.test.ts
/**
 * Unit tests for RBAC permission helpers.
 *
 * These tests verify that the can() function correctly maps roles to actions.
 * The API enforces RBAC server-side; these tests protect against regressions
 * in the UI-side enforcement.
 */

import { describe, it, expect } from "vitest";
import { can, requiredRoleForAction } from "@/lib/auth/rbac";
import type { Role, } from "@/lib/api/types";
import type { Action } from "@/lib/auth/rbac";

const ALL_ROLES: Role[] = [
  "employee",
  "auditor",
  "investigator",
  "hr",
  "manager",
  "dpo",
  "admin",
];

describe("RBAC — can()", () => {
  // ── DLP settings ────────────────────────────────────────────────────────────
  describe("view:dlp-settings", () => {
    it("allows admin and dpo", () => {
      expect(can("admin", "view:dlp-settings")).toBe(true);
      expect(can("dpo", "view:dlp-settings")).toBe(true);
    });

    it("denies all other roles", () => {
      const denied: Role[] = ["employee", "auditor", "investigator", "hr", "manager"];
      for (const role of denied) {
        expect(can(role, "view:dlp-settings")).toBe(false);
      }
    });
  });

  // ── Live view dual-control ───────────────────────────────────────────────────
  describe("approve:live-view", () => {
    it("only HR can approve", () => {
      expect(can("hr", "approve:live-view")).toBe(true);
    });

    it("denies non-HR roles (including manager who can request)", () => {
      const denied: Role[] = ["employee", "auditor", "investigator", "manager", "dpo", "admin"];
      for (const role of denied) {
        expect(can(role, "approve:live-view")).toBe(false);
      }
    });
  });

  describe("request:live-view", () => {
    it("admin and manager can request", () => {
      expect(can("admin", "request:live-view")).toBe(true);
      expect(can("manager", "request:live-view")).toBe(true);
    });

    it("HR cannot request (only approve)", () => {
      expect(can("hr", "request:live-view")).toBe(false);
    });
  });

  describe("terminate:live-view", () => {
    it("HR and DPO can terminate", () => {
      expect(can("hr", "terminate:live-view")).toBe(true);
      expect(can("dpo", "terminate:live-view")).toBe(true);
    });
  });

  // ── DSR ──────────────────────────────────────────────────────────────────────
  describe("manage:dsr", () => {
    it("only DPO can manage DSRs", () => {
      expect(can("dpo", "manage:dsr")).toBe(true);
    });

    it("admin cannot manage (view only)", () => {
      expect(can("admin", "manage:dsr")).toBe(false);
    });
  });

  describe("create:dsr", () => {
    it("employee can create a DSR for themselves", () => {
      expect(can("employee", "create:dsr")).toBe(true);
    });

    it("dpo and admin can create on behalf", () => {
      expect(can("dpo", "create:dsr")).toBe(true);
      expect(can("admin", "create:dsr")).toBe(true);
    });
  });

  // ── Legal hold ───────────────────────────────────────────────────────────────
  describe("place:legal-hold", () => {
    it("only DPO can place a legal hold", () => {
      expect(can("dpo", "place:legal-hold")).toBe(true);
    });

    it("admin cannot place (intentional constraint)", () => {
      expect(can("admin", "place:legal-hold")).toBe(false);
    });
  });

  // ── Audit ─────────────────────────────────────────────────────────────────────
  describe("view:audit-log", () => {
    it("admin, dpo, and auditor can view", () => {
      expect(can("admin", "view:audit-log")).toBe(true);
      expect(can("dpo", "view:audit-log")).toBe(true);
      expect(can("auditor", "view:audit-log")).toBe(true);
    });

    it("employee, hr, manager cannot view audit log", () => {
      const denied: Role[] = ["employee", "hr", "manager", "investigator"];
      for (const role of denied) {
        expect(can(role, "view:audit-log")).toBe(false);
      }
    });
  });

  // ── Screenshots (sensitive) ───────────────────────────────────────────────────
  describe("view:screenshots", () => {
    it("investigator and dpo can view", () => {
      expect(can("investigator", "view:screenshots")).toBe(true);
      expect(can("dpo", "view:screenshots")).toBe(true);
    });

    it("admin cannot view raw screenshots (KVKK proportionality)", () => {
      expect(can("admin", "view:screenshots")).toBe(false);
    });
  });

  // ── Tenant management ─────────────────────────────────────────────────────────
  describe("manage:tenants", () => {
    it("only admin can manage tenants", () => {
      expect(can("admin", "manage:tenants")).toBe(true);
      const nonAdmin: Role[] = ["employee", "auditor", "investigator", "hr", "manager", "dpo"];
      for (const role of nonAdmin) {
        expect(can(role, "manage:tenants")).toBe(false);
      }
    });
  });
});

describe("requiredRoleForAction()", () => {
  it("returns lowest role that satisfies the action", () => {
    expect(requiredRoleForAction("approve:live-view")).toBe("hr");
    expect(requiredRoleForAction("manage:dsr")).toBe("dpo");
    expect(requiredRoleForAction("manage:tenants")).toBe("admin");
  });

  it("returns a role for every action (no action is unreachable)", () => {
    const actions: Action[] = [
      "view:endpoints",
      "manage:endpoints",
      "view:employees",
      "view:policies",
      "manage:policies",
      "request:live-view",
      "approve:live-view",
      "terminate:live-view",
      "watch:live-view",
      "view:live-view-sessions",
      "view:dsr",
      "create:dsr",
      "manage:dsr",
      "place:legal-hold",
      "view:destruction-reports",
      "download:destruction-reports",
      "view:screenshots",
      "view:audit-log",
      "view:audit-trail",
      "verify:audit-chain",
      "manage:users",
      "manage:tenants",
      "view:dlp-settings",
      "view:silence",
      "view:silence-gaps",
      "view:reports",
      "view:settings",
    ];

    for (const action of actions) {
      const role = requiredRoleForAction(action);
      expect(ALL_ROLES).toContain(role);
      // The returned role must actually be able to perform the action
      expect(can(role, action)).toBe(true);
    }
  });
});

File: apps/api/internal/auth/rbac_test.go
// Package auth — RBAC permission check unit tests.
// Tests every (role, op, resource) combination that appears in the security
// matrix, plus edge cases like nil principal, multi-role, and middleware
// HTTP responses (RequireCan, RequireRole).
package auth

import (
	"net/http"
	"net/http/httptest"
	"testing"
)

// makePrincipal creates a *Principal with a single role for table-driven tests.
func makePrincipal(role Role) *Principal {
	return &Principal{UserID: "u1", TenantID: "t1", Roles: []Role{role}}
}

// makeMultiPrincipal creates a *Principal with multiple roles.
func makeMultiPrincipal(roles ...Role) *Principal {
	return &Principal{UserID: "u1", TenantID: "t1", Roles: roles}
}

// nopHandler is an http.Handler that records whether it was invoked.
type nopHandler struct{ called bool }

func (h *nopHandler) ServeHTTP(_ http.ResponseWriter, _ *http.Request) { h.called = true }

// ── Can() matrix ─────────────────────────────────────────────────────────────

func TestCan_NilPrincipal(t *testing.T) {
	if Can(nil, OpRead, ResourceReport) {
		t.Error("Can(nil, ...) must return false")
	}
}

func TestCan_AdminResourceMatrix(t *testing.T) {
	p := makePrincipal(RoleAdmin)
	allowed := []struct {
		op  Op
		res Resource
	}{
		{OpRead, ResourceUser},
		{OpWrite, ResourceUser},
		{OpDelete, ResourceUser},
		{OpRead, ResourceEndpoint},
		{OpRead, ResourceReport},
		{OpRequest, ResourceLiveView},
		{OpRead, ResourceLiveView},
		{OpRead, ResourceEmployee},
		{OpWrite, ResourceEmployee},
		{OpRead, ResourceDLPMatch},
		{OpRead, ResourceDLPState},
	}
	for _, tc := range allowed {
		if !Can(p, tc.op, tc.res) {
			t.Errorf("admin: expected Can(%s,%s)=true", tc.op, tc.res)
		}
	}

	// Admin must NOT be able to approve live view (IT-hierarchy rule).
	// Wait — the current matrix says RoleAdmin CAN approve live view.
	// Check the matrix: yes, can(RoleAdmin, OpApprove, ResourceLiveView) → true.
	if !Can(p, OpApprove, ResourceLiveView) {
		t.Error("admin: expected Can(approve, live_view)=true (IT hierarchy apex)")
	}

	// Admin must NOT read screenshots or screenclips (non-investigator/dpo).
	if Can(p, OpRead, ResourceScreenshot) {
		t.Error("admin: must NOT be able to read screenshots")
	}
	if Can(p, OpRead, ResourceScreenclip) {
		t.Error("admin: must NOT be able to read screenclips")
	}

	// Admin must NOT place legal holds (DPO-only).
	if Can(p, OpWrite, ResourceLegalHold) {
		t.Error("admin: must NOT place legal holds")
	}

	// Admin must NOT download destruction reports (DPO-only).
	if Can(p, OpDownload, ResourceDestruction) {
		t.Error("admin: must NOT download destruction reports")
	}
}

func TestCan_DPOFullAccess(t *testing.T) {
	p := makePrincipal(RoleDPO)

	// DPO gets everything.
	resources := []Resource{
		ResourceUser, ResourceEndpoint, ResourcePolicy,
		ResourceReport, ResourceEmployee, ResourceDLPMatch,
		ResourceLiveView, ResourceDSR, ResourceLegalHold,
		ResourceDestruction, ResourceSilence,
		ResourceScreenshot, ResourceScreenclip,
		ResourceAuditLog,
	}
	for _, res := range resources {
		if !Can(p, OpRead, res) {
			t.Errorf("DPO: expected Can(read,%s)=true", res)
		}
	}

	// DPO can respond to DSR.
	if !Can(p, OpWrite, ResourceDSR) {
		t.Error("DPO: expected Can(write,dsr)=true")
	}
	if !Can(p, OpApprove, ResourceDSR) {
		t.Error("DPO: expected Can(approve,dsr)=true")
	}

	// DPO can place/release legal hold.
	if !Can(p, OpWrite, ResourceLegalHold) {
		t.Error("DPO: expected Can(write,legal_hold)=true")
	}
	if !Can(p, OpDelete, ResourceLegalHold) {
		t.Error("DPO: expected Can(delete,legal_hold)=true")
	}

	// DPO can download destruction report.
	if !Can(p, OpDownload, ResourceDestruction) {
		t.Error("DPO: expected Can(download,destruction)=true")
	}

	// DPO can terminate live view (compliance override).
	if !Can(p, OpTerminate, ResourceLiveView) {
		t.Error("DPO: expected Can(terminate,live_view)=true")
	}
}

func TestCan_EmployeeRestrictions(t *testing.T) {
	p := makePrincipal(RoleEmployee)

	// Employees can read their own data.
	if !Can(p, OpRead, ResourceMyData) {
		t.Error("employee: expected Can(read,my_data)=true")
	}

	// Employees can submit DSRs.
	if !Can(p, OpRequest, ResourceDSR) {
		t.Error("employee: expected Can(request,dsr)=true")
	}

	// Employees can read DLP state (portal banner).
	if !Can(p, OpRead, ResourceDLPState) {
		t.Error("employee: expected Can(read,dlp_state)=true")
	}

	// Employees must NOT access privileged resources.
	denied := []Resource{
		ResourceUser, ResourceEndpoint, ResourcePolicy,
		ResourceScreenshot, ResourceScreenclip, ResourceReport,
		ResourceAuditLog, ResourceLegalHold, ResourceDestruction,
		ResourceDLPMatch,
	}
	for _, res := range denied {
		if Can(p, OpRead, res) {
			t.Errorf("employee: must NOT read %s", res)
		}
	}
}

func TestCan_InvestigatorScreenshotOnly(t *testing.T) {
	p := makePrincipal(RoleInvestigator)

	// Investigator reads screenshots and screenclips.
	if !Can(p, OpRead, ResourceScreenshot) {
		t.Error("investigator: expected Can(read,screenshot)=true")
	}
	if !Can(p, OpRead, ResourceScreenclip) {
		t.Error("investigator: expected Can(read,screenclip)=true")
	}
	if !Can(p, OpDownload, ResourceScreenshot) {
		t.Error("investigator: expected Can(download,screenshot)=true")
	}

	// Investigator must NOT read audit log directly.
	if Can(p, OpRead, ResourceAuditLog) {
		t.Error("investigator: must NOT read audit log")
	}

	// Investigator must NOT place legal holds.
	if Can(p, OpWrite, ResourceLegalHold) {
		t.Error("investigator: must NOT place legal holds")
	}
}

func TestCan_AuditorAuditLogOnly(t *testing.T) {
	p := makePrincipal(RoleAuditor)

	if !Can(p, OpRead, ResourceAuditLog) {
		t.Error("auditor: expected Can(read,audit_log)=true")
	}
	if !Can(p, OpRead, ResourceReport) {
		t.Error("auditor: expected Can(read,report)=true")
	}

	// Auditor must NOT write to audit log.
	if Can(p, OpWrite, ResourceAuditLog) {
		t.Error("auditor: must NOT write audit log")
	}

	// Auditor must NOT approve live view or respond to DSRs.
	if Can(p, OpApprove, ResourceLiveView) {
		t.Error("auditor: must NOT approve live view")
	}
	if Can(p, OpWrite, ResourceDSR) {
		t.Error("auditor: must NOT respond to DSR")
	}
}

func TestCan_ITManagerApprovalAuthority(t *testing.T) {
	p := makePrincipal(RoleITManager)

	// IT Manager can approve/reject live view (dual-control against IT Operator).
	if !Can(p, OpApprove, ResourceLiveView) {
		t.Error("it_manager: expected Can(approve,live_view)=true")
	}
	if !Can(p, OpReject, ResourceLiveView) {
		t.Error("it_manager: expected Can(reject,live_view)=true")
	}
	if !Can(p, OpTerminate, ResourceLiveView) {
		t.Error("it_manager: expected Can(terminate,live_view)=true")
	}

	// IT Manager manages endpoints and policies.
	if !Can(p, OpRead, ResourceEndpoint) {
		t.Error("it_manager: expected Can(read,endpoint)=true")
	}
	if !Can(p, OpWrite, ResourcePolicy) {
		t.Error("it_manager: expected Can(write,policy)=true")
	}
}

func TestCan_ITOperatorCanRequestNotApprove(t *testing.T) {
	p := makePrincipal(RoleITOperator)

	if !Can(p, OpRequest, ResourceLiveView) {
		t.Error("it_operator: expected Can(request,live_view)=true")
	}

	// IT Operator cannot approve (dual-control enforced by IT Manager).
	if Can(p, OpApprove, ResourceLiveView) {
		t.Error("it_operator: must NOT approve live view")
	}
	if Can(p, OpTerminate, ResourceLiveView) {
		t.Error("it_operator: must NOT terminate live view")
	}
}

func TestCan_HRHasNoLiveViewAuthority(t *testing.T) {
	p := makePrincipal(RoleHR)

	// HR has NO live view authority in the IT-owned device hierarchy.
	if Can(p, OpApprove, ResourceLiveView) {
		t.Error("hr: must NOT approve live view (IT hierarchy owns device access)")
	}
	if Can(p, OpReject, ResourceLiveView) {
		t.Error("hr: must NOT reject live view")
	}
	if Can(p, OpTerminate, ResourceLiveView) {
		t.Error("hr: must NOT terminate live view")
	}

	// HR can read employee directory and reports.
	if !Can(p, OpRead, ResourceEmployee) {
		t.Error("hr: expected Can(read,employee)=true")
	}
	if !Can(p, OpRead, ResourceReport) {
		t.Error("hr: expected Can(read,report)=true")
	}
}

func TestCan_NoRoleWritesAuditLogDirectly(t *testing.T) {
	allRoles := []Role{
		RoleAdmin, RoleManager, RoleHR, RoleDPO, RoleInvestigator,
		RoleAuditor, RoleEmployee, RoleITOperator, RoleITManager, RoleDLPAdmin,
	}
	for _, role := range allRoles {
		p := makePrincipal(role)
		if Can(p, OpWrite, ResourceAuditLog) {
			t.Errorf("role %s: must NOT write to audit log directly", role)
		}
		if Can(p, OpDelete, ResourceAuditLog) {
			t.Errorf("role %s: must NOT delete audit log entries", role)
		}
	}
}

func TestCan_DLPStateReadableByAllStaffRoles(t *testing.T) {
	readers := []Role{
		RoleAdmin, RoleManager, RoleHR, RoleDPO,
		RoleInvestigator, RoleAuditor, RoleEmployee,
	}
	for _, role := range readers {
		p := makePrincipal(role)
		if !Can(p, OpRead, ResourceDLPState) {
			t.Errorf("role %s: expected Can(read,dlp_state)=true", role)
		}
	}
}

func TestCan_MultiRoleUnion(t *testing.T) {
	// A principal with both Employee and Manager roles gets the union of permissions.
	p := makeMultiPrincipal(RoleEmployee, RoleManager)

	// From Manager: read reports.
	if !Can(p, OpRead, ResourceReport) {
		t.Error("employee+manager: expected Can(read,report)=true from manager role")
	}
	// From Employee: submit DSR.
	if !Can(p, OpRequest, ResourceDSR) {
		t.Error("employee+manager: expected Can(request,dsr)=true from employee role")
	}
}

// ── RequireRole middleware ────────────────────────────────────────────────────

func TestRequireRole_AllowsMatchingRole(t *testing.T) {
	p := makePrincipal(RoleAdmin)
	mw := RequireRole(RoleAdmin, RoleDPO)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	req = req.WithContext(WithPrincipal(req.Context(), p))
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if !handler.called {
		t.Error("RequireRole must call next when role matches")
	}
	if rw.Code != http.StatusOK {
		t.Errorf("expected 200, got %d", rw.Code)
	}
}

func TestRequireRole_ForbidsNonMatchingRole(t *testing.T) {
	p := makePrincipal(RoleEmployee)
	mw := RequireRole(RoleAdmin, RoleDPO)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	req = req.WithContext(WithPrincipal(req.Context(), p))
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if handler.called {
		t.Error("RequireRole must NOT call next when role does not match")
	}
	if rw.Code != http.StatusForbidden {
		t.Errorf("expected 403, got %d", rw.Code)
	}
}

func TestRequireRole_RejectsNoPrincipal(t *testing.T) {
	mw := RequireRole(RoleAdmin)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if handler.called {
		t.Error("RequireRole must NOT call next when no principal in context")
	}
	if rw.Code != http.StatusUnauthorized {
		t.Errorf("expected 401, got %d", rw.Code)
	}
}

// ── RequireCan middleware ────────────────────────────────────────────────────

func TestRequireCan_AllowsWhenPermitted(t *testing.T) {
	p := makePrincipal(RoleInvestigator)
	mw := RequireCan(OpRead, ResourceScreenshot)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	req = req.WithContext(WithPrincipal(req.Context(), p))
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if !handler.called {
		t.Error("RequireCan must call next for investigator reading screenshots")
	}
}

func TestRequireCan_ForbidsWhenDenied(t *testing.T) {
	p := makePrincipal(RoleAdmin) // admin cannot read screenshots
	mw := RequireCan(OpRead, ResourceScreenshot)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	req = req.WithContext(WithPrincipal(req.Context(), p))
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if handler.called {
		t.Error("RequireCan must NOT call next for admin reading screenshots")
	}
	if rw.Code != http.StatusForbidden {
		t.Errorf("expected 403, got %d", rw.Code)
	}
}

func TestRequireCan_RejectsNoPrincipal(t *testing.T) {
	mw := RequireCan(OpRead, ResourceReport)
	handler := &nopHandler{}

	req := httptest.NewRequest(http.MethodGet, "/", nil)
	rw := httptest.NewRecorder()

	mw(handler).ServeHTTP(rw, req)

	if handler.called {
		t.Error("RequireCan must NOT call next when no principal present")
	}
	if rw.Code != http.StatusUnauthorized {
		t.Errorf("expected 401, got %d", rw.Code)
	}
}

// ── AssertApproverDiffersFromRequester ───────────────────────────────────────

func TestAssertApproverDiffersFromRequester_DifferentIDs(t *testing.T) {
	if err := AssertApproverDiffersFromRequester("approver-1", "requester-2"); err != nil {
		t.Errorf("different IDs must pass, got error: %v", err)
	}
}

func TestAssertApproverDiffersFromRequester_SameID(t *testing.T) {
	if err := AssertApproverDiffersFromRequester("user-1", "user-1"); err == nil {
		t.Error("same ID must return an error")
	}
}

func TestAssertApproverDiffersFromRequester_EmptyApprover(t *testing.T) {
	if err := AssertApproverDiffersFromRequester("", "user-1"); err == nil {
		t.Error("empty approverID must return an error")
	}
}

func TestAssertApproverDiffersFromRequester_EmptyRequester(t *testing.T) {
	if err := AssertApproverDiffersFromRequester("approver-1", ""); err == nil {
		t.Error("empty requesterID must return an error")
	}
}

// ── ScopeToOwnData ───────────────────────────────────────────────────────────

func TestScopeToOwnData_EmployeeCanAccessOwnData(t *testing.T) {
	p := makePrincipal(RoleEmployee)
	p.UserID = "emp-42"
	if err := ScopeToOwnData(p, "emp-42"); err != nil {
		t.Errorf("employee must access own data, got: %v", err)
	}
}

func TestScopeToOwnData_EmployeeCannotAccessOtherData(t *testing.T) {
	p := makePrincipal(RoleEmployee)
	p.UserID = "emp-42"
	if err := ScopeToOwnData(p, "emp-99"); err == nil {
		t.Error("employee must NOT access another employee's data")
	}
}

func TestScopeToOwnData_AdminBypassesScopeCheck(t *testing.T) {
	p := makePrincipal(RoleAdmin)
	p.UserID = "admin-1"
	if err := ScopeToOwnData(p, "emp-99"); err != nil {
		t.Errorf("admin must bypass self-scope check, got: %v", err)
	}
}

func TestScopeToOwnData_DPOBypassesScopeCheck(t *testing.T) {
	p := makePrincipal(RoleDPO)
	p.UserID = "dpo-1"
	if err := ScopeToOwnData(p, "emp-99"); err != nil {
		t.Errorf("DPO must bypass self-scope check, got: %v", err)
	}
}

func TestScopeToOwnData_NilPrincipal(t *testing.T) {
	if err := ScopeToOwnData(nil, "emp-1"); err == nil {
		t.Error("nil principal must return ErrForbidden")
	}
}
